recipe/meta.yaml

```diff
@@ -25,26 +22,28 @@ build:
 
 requirements:
   build:
-    - cairo 1.14.*             # [linux]
-    - expat 2.2.*                   # [linux]
-    - libpng >=1.6.32,<1.6.35  # [unix]
-    - libtiff >=4.0.8,<4.0.10   # [unix]
-    - libtool                  # [linux]
-    - freetype 2.8.1             # [unix]
-    - jpeg 9*                  # [unix]
-    - pango 1.40.*             # [linux]
+    - {{ compiler('cxx') }}
     - pkg-config               # [osx]
-    - zlib 1.2.11               # [unix]
+  host:
+    - cairo                    # [linux]
+    - expat                    # [linux]
+    - libpng                   # [unix]
+    - libtiff                  # [unix]
+    - libtool                  # [linux]
+    - freetype                 # [unix]
+    - jpeg                     # [unix]
+    - pango                    # [linux]
+    - zlib                     # [unix]
   run:
-    - cairo 1.14.*             # [linux]
-    - expat 2.2.*                   # [linux]
-    - libpng >=1.6.32,<1.6.35  # [unix]
-    - libtiff >=4.0.8,<4.0.10   # [unix]
+    - cairo                    # [linux]
+    - expat                    # [linux]
+    - libpng                   # [unix]
+    - libtiff                  # [unix]
     - libtool                  # [linux]
-    - freetype 2.8.1             # [unix]
-    - jpeg 9*                  # [unix]
-    - pango 1.40.*             # [linux]
-    - zlib 1.2.11               # [unix]
+    - freetype                 # [unix]
+    - jpeg                     # [unix]
+    - pango                    # [linux]
+    - zlib                     # [unix]
```

Comment:
This is really bad! Please restore all the pinnings here.

Why did you remove them?

Reply:
I removed them because we have the same pinnings in the pinning feedstock.

Reply:
> in the sense that it would be incompatible with other versions

That's the main reason those pinnings are in place, e.g. to avoid incompatibilities with different versions of freetype, jpeg, etc.

@jakirkham?

Reply:
What I mean is: if a package needs a certain version because of the API, it should be pinned in the dependencies. If a package needs to pin just to maintain ABI compatibility (the runtime version has to be compatible to the build time version) this can be handled by CB3 via the pinning feedstock.

Reply:
By the pinning feedstock, it would even be possible to build grahpviz for different versions of e.g. freetype (which we don't do though). Also, updating the pinning in the feedstock means that all packages can be consistently updated to a new pinning, such that it is easier to combine them in the same environment.

Reply:
Also see [the docs](https://conda.io/docs/user-guide/tasks/build-packages/variants.html)

Reply:
Johannes is correct.

We migrated what was in the old pinning script to that pinning file. If you look at the dependencies here and that pinning file, they should be pretty close (if not exact) matches to what we have here. A few shifted slightly over time to address issues that came up in the beginning.

Re-rendering pulls a subset of those pinnings into the feedstock (just did that now). So one can see how these will be pinned and verify they are correct.

Happy to answer any questions. :)

Reply:
Ok, I see, thanks for the clarification @johanneskoester and @jakirkham.

But this is a feature of conda-smithy 3, paired with conda-build 3, right? I mean, it sounded strange to me when @johanneskoester said that pinnings are auto-magically handled by conda-build only.

Reply:
Sure there is some work being done in the background by our tooling.

In the future, we may move to [`run_exports`]( https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#export-runtime-requirements ), which would avoid needing this work done by `conda-smithy`. Example usage can be seen with [`zlib`]( https://github.com/conda-forge/zlib-feedstock/blob/fa4368510c67a2dd14c19e0ed2791bd53ed958bb/recipe/meta.yaml#L14-L17 ).